service-level-indicators/spark-retrieval-success-rate.md

```diff
@@ -0,0 +1,612 @@
+
+As previously mentioned, the checkers start each round by downloading the Round Retrieval Task List. For each task that’s in the List, the checker calculates the task’s “key” by SHA256-hashing the task and the current dRand randomness, fetched from the [dRand network API](https://github.com/filecoin-station/spark-evaluate/blob/a231822d3d78e3d096425a53a300f8c6c82ee01f/lib/drand-client.js#L29-L33). Leveraging the wonderful properties of cryptographic hash functions, these hash values will be randomly & uniformly distributed in the entire space of $2^{256}$ values. Also, by including the randomness in the input alongside the Eligible Deal details, we will get a different hash digest for a given Eligible Deal in each round. We can define the `taskKey` as
+
+```jsx
+taskKey = SHA256(payloadCid || miner_id || drand)
+```
+
+Each Spark Checker node can also calculate the SHA256 hash of its Station Id. This is fixed across rounds as it doesn’t depend on any round-specific inputs. (In the future, the protocol may add the drand value into the hash input of the nodeKey too.)
+
+```jsx
+nodeKey = SHA256(station_id)
+```
+
+The checker can then find its $k$ “closest” tasks, using XOR as the distance metric. These $k$ tasks are the Retrieval Tasks the Spark Checker node is eligible to complete.  Any other tasks submitted by the checker are not evaluated or rewarded.
+
+```jsx
+dist = taskKey XOR nodeKey
+```
+
+Note that, at the start of each round, the protocol doesn’t know which Spark Checkers will participate as there are no uptime requirements on these checkers. This means the Spark protocol can’t centrally form groups of checkers and assign them a subset of the Round Retrieval Task List. The above approach doesn’t make any assumptions about which checkers are online, but instead relies on the fact that the `nodeKeys` will be evenly distributed around the SHA256 hash ring.
+
+Following the above approach, for each Retrieval Task, there are a set of Checkers who find this task among their $k$ closest and they will attempt the task in the round. We refer to the set of checkers who attempt a Retrieval Task as the “[Committee](#committee) ” for that task. Since we are hashing over a random value for each task when we create its `taskKey`, this approach mitigates against one party controlling an entire committee and dominating the honest majority consensus decision later in the protocol.
+
+### **How many tasks should each checker do per round? What value is given to $k$?**
+
+The choice of $k$ is determined by Spark protocol logic that aims to keep the overall number of Spark measurements completed by the total network per round fixed. This is important because we don’t want the number of requests that Storage Providers need to deal with go up as the number of Spark Checkers in the network increases.
```

Comment:
```suggestion
The choice of $k$ is determined by Spark protocol logic that aims to keep the overall number of Spark measurements completed by the total network per round fixed.

This is important because
- We don’t want the number of requests that Storage Providers need to deal with go up as the number of Spark Checkers in the network increases
- There need to be enough nodes in each committee for its result to be considered reliable
```

Reply:
```suggestion
The choice of $k$ is determined by Spark protocol logic that aims to keep the overall number of Spark measurements completed by the total network per round fixed.
This is important because
- We don’t want the number of requests that Storage Providers need to deal with to go up as the number of Spark Checkers in the network increases
- There needs to be enough nodes in each committee for its result to be considered reliable
```